src/vm/checker/typecheck/mod.rs

```diff
@@ -0,0 +1,404 @@
+pub enum FunctionType {
+    Variadic(TypeSignature, TypeSignature),
+    Fixed(Vec<TypeSignature>, TypeSignature)
+}
+
+pub struct TypeChecker <'a> {
+    pub type_map: TypeMap,
+    contract_context: ContractContext,
+    db: &'a AnalysisDatabase
+}
+
+impl FunctionType {
+    pub fn check_args(&self, args: &[TypeSignature]) -> CheckResult<()> {
+        match self {
+            FunctionType::Variadic(expected_type, _) => {
+                if args.len() < 1 {
+                    return Err(CheckError::new(CheckErrors::VariadicNeedsOneArgument))
+                }
+                for found_type in args.iter() {
+                    if !expected_type.admits_type(found_type) {
+                        return Err(CheckError::new(CheckErrors::TypeError(
+                            expected_type.clone(), found_type.clone())))
+                    }                    
+                }
+                Ok(())
+            },
+            FunctionType::Fixed(arg_types, _) => {
+                if arg_types.len() != args.len() {
+                    return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(
+                        arg_types.len(), args.len())))
+                }
+                for (expected_type, found_type) in arg_types.iter().zip(args) {
+                    if !expected_type.admits_type(found_type) {
+                        return Err(CheckError::new(CheckErrors::TypeError(
+                            expected_type.clone(), found_type.clone())))
+                    }
+                }
+                Ok(())
+            }
+        }
+    }
+
+    pub fn return_type(&self) -> TypeSignature {
+        match self {
+            FunctionType::Variadic(_, return_type) => return_type.clone(),
+            FunctionType::Fixed(_, return_type) => return_type.clone()
+        }
+    }
+}
+
+fn type_reserved_variable(variable_name: &str) -> Option<TypeSignature> {
+    if let Some(variable) = NativeVariables::lookup_by_name(variable_name) {
+        use vm::variables::NativeVariables::*;
+        let var_type = match variable {
+            TxSender => TypeSignature::new_atom(AtomTypeIdentifier::PrincipalType),
+            BlockHeight => TypeSignature::new_atom(AtomTypeIdentifier::IntType),
+            BurnBlockHeight => TypeSignature::new_atom(AtomTypeIdentifier::IntType),
+        };
+        Some(var_type)
+    } else {
+        None
+    }
+}
+
+fn no_type() -> TypeSignature {
+    TypeSignature::new_atom(AtomTypeIdentifier::NoType)
+}
+
+fn check_atomic_type(atom: AtomTypeIdentifier, to_check: &TypeSignature) -> CheckResult<()> {
+    let expected = TypeSignature::new_atom(atom);
+    if !expected.admits_type(to_check) {
+        Err(CheckError::new(CheckErrors::TypeError(expected, to_check.clone())))
+    } else {
+        Ok(())
+    }
+}
+
+impl <'a> TypeChecker <'a> {
+    pub fn new(db: &'a AnalysisDatabase) -> TypeChecker {
+        TypeChecker {
+            db: db,
```

Comment:
Just to be clear -- the database is _not_ modified while a contract is being type-checked, right?  Except, that is, to insert the final analysis itself.

Reply:
Yep -- during TypeCheck, it is only ever used for looking up types from other contracts.